### PR TITLE
feat(adapter-nextjs): add createAuthRouteHandlers to createServerRunner

### PR DESCRIPTION
**Pull request by HuiSF**
#### Checklist

- [ ] PR description included
- [ ] `yarn test` passes
- [ ] Unit Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)
- [ ] Relevant documentation is changed or added (and PR referenced)

#### Checklist for repo maintainers

- [ ] Verify E2E tests for existing workflows are working as expected or add E2E tests for newly added workflows
- [ ] New source file paths included in this PR have been added to CODEOWNERS, if appropriate

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/adapter-nextjs/src/auth/utils/handlerParametersTypeAssertions.ts [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+}
+
+// NextApiRequest is the 1st parameter type for the API route handlers in the Pages Router
+export function isNextApiRequest(request: object): request is NextApiRequest {
+	// Can't use `IncomingMessage` to validate the request is an instance of `NextApiRequest`
+	// as `import from 'http'` breaks the Next.js build.
+	// The `query` property is a convenience method added to the underlying `IncomingMessage`.
+	return (
+		'query' in request &&
+		Object.prototype.toString.call(request.query) === '[object Object]'
```
Comment (jjarvisp) [inline packages/adapter-nextjs/src/auth/utils/handlerParametersTypeAssertions.ts:41]: Any particular reason to deviate from the typesafe object check on 22-23?

**Review by jjarvisp — APPROVED** [review]
Looks good to me. Unit tests are thorough.

File: packages/adapter-nextjs/__tests__/auth/utils/isSupportedAuthApiRoutePath.test.ts [hunk 2]
```diff
@@ -0,0 +1,14 @@
+import { isSupportedAuthApiRoutePath } from '../../../src/auth/utils/isSupportedAuthApiRoutePath';
+
+describe('isSupportedAuthApiRoutePath', () => {
+	test.each([
```
Comment (cshfang) [inline packages/adapter-nextjs/__tests__/auth/utils/isSupportedAuthApiRoutePath.test.ts:4]: Nit: Do you need/want to test for the undefined case?

File: packages/adapter-nextjs/__tests__/auth/handleAuthApiRouteRequestForPagesRouter.test.ts [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+		redirectSignIn: ['https://example.com/signin'],
+		redirectSignOut: ['https://example.com/signout'],
+		responseType: 'code',
+		scopes: ['openid', 'email'],
+	};
+	const testSetCookieOptions = {};
+
+	it('sets response.status(405) when request has an unsupported method', () => {
+		const mockEnd = jest.fn();
+		const mockStatus = jest.fn(() => ({ end: mockEnd }));
```
Comment (cshfang) [inline packages/adapter-nextjs/__tests__/auth/handleAuthApiRouteRequestForPagesRouter.test.ts:23]: Nit: These look like they can be extracted and just reused with mock clears? Would help reduce some boilerplate per test to help readability

File: packages/adapter-nextjs/__tests__/auth/utils/handlerParametersTypeAssertions.test.ts [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+			false,
+		],
+		[
+			{
+				params: {
+					slug: 'sign-in',
+				},
+			},
+			true,
+		],
```
Comment (cshfang) [inline packages/adapter-nextjs/__tests__/auth/utils/handlerParametersTypeAssertions.test.ts:29]: Nit: Could these really not fit on single lines? This is kinda rough to read

File: packages/adapter-nextjs/src/auth/utils/isSupportedAuthApiRoutePath.ts [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+// SPDX-License-Identifier: Apache-2.0
+
+import { supportedRoutePaths } from '../constant';
+import { SupportedRoutePaths } from '../types';
+
+export function isSupportedAuthApiRoutePath(
+	path?: string,
+): path is SupportedRoutePaths {
+	return (
+		path !== undefined &&
```
[the remark this answers is not in the document]
Comment (HuiSF) [reply packages/adapter-nextjs/src/auth/utils/isSupportedAuthApiRoutePath.ts:11]: ts doesn't like `supportedRoutePaths.includes(undefined)` as `undefined` is a not allow element type in `SupportedRoutePaths`

File: packages/adapter-nextjs/src/auth/utils/handlerParametersTypeAssertions.ts (hunk 1 above)
Comment (HuiSF) [reply packages/adapter-nextjs/src/auth/utils/handlerParametersTypeAssertions.ts:41]: This is specifically for checking whether `query` is a key-value map - but not other type of object.

File: packages/adapter-nextjs/src/auth/createAuthRouteHandlersFactory.ts [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+	isNextRequest,
+} from './utils';
+import { handleAuthApiRouteRequestForAppRouter } from './handleAuthApiRouteRequestForAppRouter';
+import { handleAuthApiRouteRequestForPagesRouter } from './handleAuthApiRouteRequestForPagesRouter';
+
+export const createAuthRouteHandlersFactory = ({
+	config: resourcesConfig,
+	runtimeOptions = {},
+}: CreateAuthRouteHandlersFactoryInput): CreateAuthRouteHandlers => {
+	const origin = process.env.AMPLIFY_APP_ORIGIN;
```
[the remark this answers is not in the document]
Comment (HuiSF) [reply packages/adapter-nextjs/src/auth/createAuthRouteHandlersFactory.ts:31]: This is a good call, can move up getting the origin to the root factory to keep this function pure.

File: packages/adapter-nextjs/src/auth/utils/handlerParametersTypeAssertions.ts [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+	// NextRequest extends the Web Request API with additional convenience methods.
+	// Details: https://nextjs.org/docs/app/api-reference/functions/next-request#nexturl
+	return request instanceof Request && 'nextUrl' in request;
+}
+
+// AuthRoutesHandlersContext is the 2nd parameter type for the API route handlers in the App Router
+export function isAuthRoutesHandlersContext(
+	context: object,
+): context is AuthRoutesHandlerContext {
+	return (
```
[the remark this answers is not in the document]
Comment (HuiSF) [reply packages/adapter-nextjs/src/auth/utils/handlerParametersTypeAssertions.ts:20]: This one in particular to narrow the types down to make each step can pass TS's check.

File: packages/adapter-nextjs/src/auth/utils/predicates.ts [hunk 8]
```diff
@@ -19,7 +19,7 @@ export function isAuthRoutesHandlersContext(
 ): context is AuthRoutesHandlerContext {
 	return (
 		'params' in context &&
-		typeof context.params === 'object' &&
+		context.params !== undefined &&
```
Comment (cshfang) [inline packages/adapter-nextjs/src/auth/utils/predicates.ts:22]: Nit: save a line?
```
!= null
```

**Review by cshfang — APPROVED** [review]

File: packages/adapter-nextjs/src/auth/utils/predicates.ts (hunk 8 above)
Comment (jjarvisp) [reply packages/adapter-nextjs/src/auth/utils/predicates.ts:22]: I get the motivation but personally would prefer extra line over non-strict equals; it's arguably less intuitive for future maintainers

**Review by jjarvisp — APPROVED** [review]

**Pull request merged**